cirq-google/cirq_google/api/v2/results.py

```diff
@@ -54,34 +50,31 @@ def find_measurements(program: cirq.AbstractCircuit) -> List[MeasureInfo]:
         NotImplementedError: If the program is of a type that is not recognized.
         ValueError: If there is a duplicate measurement key.
     """
-    measurements: List[MeasureInfo] = []
-    keys: Set[str] = set()
-
-    if isinstance(program, cirq.AbstractCircuit):
-        measure_iter = _circuit_measurements(program)
-    else:
+    if not isinstance(program, cirq.AbstractCircuit):
         raise NotImplementedError(f'Unrecognized program type: {type(program)}')
 
-    for m in measure_iter:
-        if m.key in keys:
-            raise ValueError(f'Duplicate measurement key: {m.key}')
-        keys.add(m.key)
-        measurements.append(m)
-
-    return measurements
-
-
-def _circuit_measurements(circuit: cirq.AbstractCircuit) -> Iterator[MeasureInfo]:
-    for i, moment in enumerate(circuit):
+    measurements: Dict[str, MeasureInfo] = {}
+    instances: Dict[str, int] = Counter()
+    for moment in program:
         for op in moment:
             if isinstance(op.gate, cirq.MeasurementGate):
-                yield MeasureInfo(
+                m = MeasureInfo(
                     key=op.gate.key,
                     qubits=_grid_qubits(op),
-                    slot=i,
+                    instances=1,
```

Comment:
I find this construction a little weird and I think it's mostly because of this `instances=1` parameter. I see why you need it to do the comparison, but it's also bizarre to not update the `instances` count that we just set. A couple of ideas:

1. Override the equality operator on `MeasureInfo` to ignore `instances` and then increment the value in the dict if it already exists.
2. Make `instances` an optional parameter with default value 1, that way you're not calling attention to it here but eventually assign it later. This could be combined with #1 to make it more clear that you're really just comparing the other values.

You may have better ideas for this, but I think it could use a little work. WDYT?

Reply:
Reworked this to update the `instances` count as we go, and explicitly check the other attributes that need to be compatible.